libcore/src/events/EventManager.cpp

```diff
@@ -38,74 +38,42 @@ EventManager::EventManager(Building * _b)
     _building = _b;
 }
 
-void EventManager::AddEvents(std::vector<Event> events)
+void EventManager::AddEvent(std::unique_ptr<Event> event)
 {
-    _events.insert(_events.end(), events.begin(), events.end());
+    event->SetBuilding(_building);
+    _events.emplace_back(std::move(event));
+    std::sort(std::begin(_events), std::end(_events), [](auto & event1, auto & event2) {
+        return event1->GetTime() < event2->GetTime();
+    });
 }
 
 void EventManager::ListEvents()
 {
     for(const auto & event : _events) {
-        LOG_INFO("{}", event.GetDescription());
+        LOG_INFO("{}", event->GetDescription());
     }
 }
 
 bool EventManager::ProcessEvent()
 {
-    bool eventProcessed = false;
+    double timeMin = Pedestrian::GetGlobalTime() - J_EPS_EVENT;
+    std::unique_ptr<Event> eventMin =
+        std::make_unique<DoorEvent>(-1, timeMin, EventAction::DOOR_OPEN);
 
-    for(const auto & event : _events) {
-        if(fabs(event.GetTime() - Pedestrian::GetGlobalTime()) < J_EPS_EVENT) {
-            //Event with current time stamp detected
-            LOG_INFO("Event: after {:.2f} sec", Pedestrian::GetGlobalTime());
-            switch(event.GetAction()) {
-                case EventAction::OPEN:
-                    OpenDoor(event.GetId());
-                    break;
-                case EventAction::CLOSE:
-                    CloseDoor(event.GetId());
-                    break;
-                case EventAction::TEMP_CLOSE:
-                    TempCloseDoor(event.GetId());
-                    break;
-                case EventAction::RESET_USAGE:
-                    ResetDoor(event.GetId());
-                    break;
-                case EventAction::NOTHING:
-                    LOG_WARNING("Unknown event action in events. open, close, reset or "
-                                "temp_close. Default: do nothing");
-                    break;
-            }
-            eventProcessed = true;
-        }
-    }
-    return eventProcessed;
-}
+    auto firstEventAtTime = std::upper_bound(
+        std::begin(_events), std::end(_events), eventMin, [](auto & val, auto & event) {
```

Comment:
Maybe we should implement a lower operator on Event? Would clean up this part, but maybe it is not clear what lower means on event?

Reply:
Unfortunately comparison operators won't work as unique_ptr are saved in the container, which would lead to comparison of the pointers and not the content.